metrics.nim

```diff
@@ -938,16 +938,26 @@ when defined(metrics):
       # TODO: parse the output of `GC_getStatistics()` for more stats
 
       when defined(nimTypeNames) and declared(dumpHeapInstances):
-        var heapSizes: array[10, (cstring, int)]
+        # Higher size then in the loop for adding metrics
+        # to avoid missing same name metrics far apart with low values.
+        var heapSizes: array[100, (cstring, int)]
+        var counter: int
         for data in dumpHeapInstances():
+          counter += 1
           var smallest = 0
-          for i in 1..<heapSizes.len:
+          var dedupe = false
+          for i in 0..<heapSizes.len:
+            if heapSizes[i][0] == data.name:
+              heapSizes[i][1] += data.sizes
+              dedupe = true
+              break
             if heapSizes[smallest][1] >= heapSizes[i][1]:
               smallest = i
-          if data.sizes > heapSizes[smallest][1]:
+          if not dedupe and data.sizes > heapSizes[smallest][1]:
             heapSizes[smallest] = (data.name, data.sizes)
         sort(heapSizes, proc(a, b: auto): auto = b[1] - a[1])
-        for (typeName, size) in heapSizes:
+        # Lower the number of metrics to reduce metric cardinality.
+        for (typeName, size) in heapSizes[..(heapSizes.len/10)]:
```

Comment:
```suggestion
        for i in 0..<10:
          let (typeName, size) = heapSizes[i]
```

a few more nim curiosities:
* `[..]` allocates a new sequence unnecessarily 
* `/` returns a float - one can use `heapSizes.len div 10` to get an integer

I'm surprised this compiles at all - alternatively, it's simply not covered by unit tests

Reply:
Wasn't aware that a slice allocates memory. But shouldn't that be an issue only within the loop that iterates over values from `dumpHeapInstances`. This slice happens after that loop exits.

Reply:
it's not wrong per se - it's just wasteful (the gc then has to deallocate the memory - ie we avoid it when we can.

I'm more surprised that the range with a float in it works :O

Reply:
the way to test whether it compiles for example this is to update the submodule of some project like nimbus-eth2 that uses metrics to this particular commit, then compile that project

Reply:
I tested this and it doesn't seem to fail:
```nim
let test: array[4, string] = [
  "a", "b", "c", "d"
]

echo test[(3/2)..(7/3)]
```
Result:
```
@["b", "c"]
```
Not sure if that's intended behavior, but I'll change it to what you suggested with `div`.

Reply:
interesting, guess I learned something today :)

Reply:
And regarding the slice. Would this be a better approach?
```nim
for i in 0..(heapSizes.len div 10):
  let typeName = heapSizes[i].typeName
  let size = heapSizes[i].size
```

Reply:
it's roughly the same as `let (typeName, size) = heapSizes[i]`. also, keep in mind that Nim slices are closed intervals, meaning inclusive - this will give you one item more than most people expect.. `..<` is used to get the normal behaviour

Reply:
whether it's `x.len div 10` or simple `10` is mostly a matter of taste and one could argue both ways, so whatever.

Reply:
I think having a variable with a static limit of `10` makes it clearer what's the intention. Added it and changed the loop.